lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -62,10 +62,68 @@ class TraceProcessor {
   }
 
   /**
+   * This method sorts a group of trace events that have the same timestamp. We want to...
+   *
+   * 1. Put E events first, we finish off our existing events before we start new ones.
+   * 2. Order B/X events by their duration, we want parents to start before child events.
+   * 3. If we don't have any of this to go on, just use the position in the original array (stable sort).
+   *
+   * @param {number[]} tsGroupIndices
+   * @param {number[]} timestampSortedIndices
+   * @param {number} indexOfTsGroupIndicesStart
+   * @param {LH.TraceEvent[]} traceEvents
+   * @return {number[]}
+   */
+  static _sortTimestampEventGroup(
+      tsGroupIndices,
+      timestampSortedIndices,
+      indexOfTsGroupIndicesStart,
+      traceEvents
+  ) {
+    const eEventIndices = tsGroupIndices.filter(index => traceEvents[index].ph === 'E');
+    const bxEventIndices = tsGroupIndices.filter(index => traceEvents[index].ph === 'X' ||
+      traceEvents[index].ph === 'B');
+    const otherEventIndices = tsGroupIndices.filter(index => !eEventIndices.includes(index) &&
+      !bxEventIndices.includes(index));
+
+    /** @type {Map<number, number>} */
+    const effectiveDuration = new Map();
+    for (const index of bxEventIndices) {
+      const event = traceEvents[index];
+      if (event.ph === 'X') {
+        effectiveDuration.set(index, event.dur);
+      } else {
+        // Find the first 'E' event *after* the current group of events that matches our name.
+        let duration = Number.MAX_SAFE_INTEGER;
+        const startIndex = indexOfTsGroupIndicesStart + tsGroupIndices.length;
+        for (let j = startIndex; j < timestampSortedIndices.length; j++) {
+          const potentialEndEvent = traceEvents[timestampSortedIndices[j]];
+          if (potentialEndEvent.ph === 'E' && potentialEndEvent.name === event.name) {
```

Comment:
with that in mind, should this also look for the matching `E` event by `pid` as well?

Reply:
> with that in mind, should this also look for the matching `E` event by `pid` as well?

while we're fixing trace processing corner cases (ONCE AND FOR ALL :), there are cases where we're actually hitting the `E` event of a *nested* event here. The effect isn't big (I'm hitting like ~3 times in a verge trace that this hits a `B` event with matching name and pid before hitting a matching `E` event), but it seems worth accounting for while we're combing through this

Reply:
> should this also look for the matching E event by pid as well?

technically we should also check `tid` as well, but in reality i dont expect that to fix anything.

Reply:
> there are cases where we're actually hitting the E event of a nested event here.

good catch! fixed 👍 i'll need to update #9491 accordingly as well

> technically we should also check tid as well, but in reality i dont expect that to fix anything.

might as well 👍